crates/core/src/operations/transaction/mod.rs

```diff
@@ -167,94 +167,155 @@ pub async fn prepare_commit<'a>(
     Ok(path)
 }
 
-/// Commit a transaction, with up to 15 retries. This is higher-level transaction API.
-///
-/// Will error early if the a concurrent transaction has already been committed
-/// and conflicts with this transaction.
-pub async fn commit(
-    log_store: &dyn LogStore,
-    actions: &Vec<Action>,
-    operation: DeltaOperation,
-    read_snapshot: Option<&DeltaTableState>,
-    app_metadata: Option<HashMap<String, Value>>,
-) -> DeltaResult<i64> {
-    commit_with_retries(
-        log_store,
-        actions,
-        operation,
-        read_snapshot,
-        app_metadata,
-        15,
-    )
-    .await
+#[derive(Default, Clone)]
+pub struct CommitProperties {
+    pub(crate) app_metadata: HashMap<String, Value>,
```

Comment:
The mixture of elements between written and configuration items annoys me too. When creating this struct the primary focus is to support transaction identifiers and eventually domain metadata which are written in the commit. Those two features should be exposed for any transaction performed on the table.

I think app metadata should be kept with the commit builder and we instead can define a merging strategy for how the final `app metadata` is created. The hierarchy can be something like `Operation Metadata` > `DeltaTableConfig Metadata` > `User Metadata`. User metadata should be last to prevent end users from overwriting critical elements like user information, notebook, etc.

Reply:
> Another conceptually larger change that I think will make our codebase a bit cleaner is to make the `Snapshot` non-optional. Even if we are creating a table we always require at least the `Protocol` and `Metadata` action, which we can use to create an empty snapshot with no file action / empty `LogSegment` and a version of `-1`.

@roeap I'm try to design the commit interface to prepare for this change. I have one question. Will `Snapshot` always contain the required information required to perform the commit? I.E Will there ever be an instance where commit will need an `EagerSnapshot` or a `DeltaTableState`? Also are there some workflows (e.g blind appends) where having a `log_segment` might not be required.

If only `Snapshot` is required then I'll change the build definition to 
```rust
pub fn build(
        self,
        snapshot: &Snapshot,
        log_store: LogStoreRef,
        operation: DeltaOperation,
    ) -> Result<PreCommit<'a>, CommitBuilderError>
```

But if different levels of information are required then we can maybe use a trait object and progressively enhance depending on what information is available. I tried playing around with this concept in the latest commit.

Interested on your thoughts and how we should proceed here.

Reply:
hmm .. `Snapshot` will always contain `Metadata` and `Protocol` - for conflict resolution we need to read new actions anyhow. so for blind appends we may not actually need to get any actions beyond that ... So not eagerly fetching all data, but catching things just in time seems like a good way to go.